test/qdmi/test_driver.cpp

```diff
@@ -0,0 +1,379 @@
+    for (auto& param : params) {
+      param = dis(gen);
+    }
+    EXPECT_EQ(QDMI_device_query_operation_property(
+                  device, op, 0, nullptr, numParams, params.data(),
+                  QDMI_OPERATION_PROPERTY_DURATION, sizeof(double), &duration,
+                  nullptr),
+              QDMI_SUCCESS)
+        << "Failed to query duration for operation " << name << ".";
+    EXPECT_EQ(QDMI_device_query_operation_property(
+                  device, op, 0, nullptr, numParams, params.data(),
+                  QDMI_OPERATION_PROPERTY_FIDELITY, sizeof(double), &fidelity,
+                  nullptr),
+              QDMI_SUCCESS)
+        << "Failed to query fidelity for operation " << name << ".";
+
+    EXPECT_EQ(QDMI_device_query_operation_property(
+                  device, op, 0, nullptr, 0, nullptr,
+                  QDMI_OPERATION_PROPERTY_MAX, 0, nullptr, nullptr),
+              QDMI_ERROR_INVALIDARGUMENT)
+        << "The MAX property is not a valid value for any device.";
+  }
+}
+
+TEST_P(DriverTest, SessionInit) {
+  EXPECT_EQ(QDMI_session_init(nullptr), QDMI_ERROR_INVALIDARGUMENT)
+      << "`session == nullptr` is not a valid argument.";
+  EXPECT_EQ(QDMI_session_init(session), QDMI_ERROR_BADSTATE)
+      << "Session must return `BADSTATE` if it is initialized again.";
+}
+
+TEST_P(DriverTest, QuerySessionProperties) {
+  EXPECT_EQ(QDMI_session_query_session_property(
+                nullptr, QDMI_SESSION_PROPERTY_DEVICES, 0, nullptr, nullptr),
+            QDMI_ERROR_INVALIDARGUMENT)
+      << "`session == nullptr` is not a valid argument.";
+  EXPECT_EQ(QDMI_session_query_session_property(
+                session, QDMI_SESSION_PROPERTY_MAX, 0, nullptr, nullptr),
+            QDMI_ERROR_INVALIDARGUMENT)
+      << "`prop >= QDMI_SESSION_PROPERTY_MAX` is not a valid argument.";
+
+  // Must not query on an uninitialized session
+  QDMI_Session uninitializedSession = nullptr;
+  ASSERT_EQ(QDMI_session_alloc(&uninitializedSession), QDMI_SUCCESS);
+  EXPECT_EQ(QDMI_session_query_session_property(uninitializedSession,
+                                                QDMI_SESSION_PROPERTY_DEVICES,
+                                                0, nullptr, nullptr),
+            QDMI_ERROR_BADSTATE);
+
+  constexpr size_t size = sizeof(QDMI_Device) - 1;
+  std::array<char, size> devices{};
+  EXPECT_EQ(QDMI_session_query_session_property(
+                session, QDMI_SESSION_PROPERTY_DEVICES, size,
+                static_cast<void*>(devices.data()), nullptr),
+            QDMI_ERROR_INVALIDARGUMENT)
+      << "Device must return `INVALIDARGUMENT` if the buffer is too small.";
+}
+
+TEST_P(DriverTest, QueryNeedsCalibration) {
+  size_t needsCalibration = 0;
+  const auto ret = QDMI_device_query_device_property(
+      device, QDMI_DEVICE_PROPERTY_NEEDSCALIBRATION, sizeof(size_t),
+      &needsCalibration, nullptr);
+  EXPECT_EQ(ret, QDMI_SUCCESS);
+  EXPECT_EQ(needsCalibration, 0);
```

Comment:
This is not an assertion we can make in general.
The only thing we can assert is that this will return 0 or 1.